src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c

```diff
@@ -0,0 +1,1041 @@
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR
+ * ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
+ * CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/*
+ * The text above constitutes the entire PortAudio license; however,
+ * the PortAudio community also makes the following non-binding requests:
+ *
+ * Any person wishing to distribute modifications to the Software is
+ * requested to send the modifications to the original developer so that
+ * they can be incorporated into the canonical version. It is also
+ * requested that these non-binding requests be included along with the
+ * license above.
+ */
+
+/** @file
+ @ingroup common_src
+
+ @brief Pulseaudio implementation of support for a host API.
+
+ This host API implements Pulseaudio support for portaudio
+ it has callbackmode and normal write mode support
+*/
+
+
+#include <string.h> /* strlen() */
+
+
+#include "pa_hostapi_pulseaudio.h"
+#include "pa_hostapi_pulseaudio_cb.h"
+#include "pa_hostapi_pulseaudio_block.h"
+
+
+/* Pulseaudio headers */
+#include <stdio.h>
+#include <string.h>
+#include <pulse/pulseaudio.h>
+
+
+/* prototypes for functions declared in this file */
+
+#define PA_PULSEAUDIO_SET_LAST_HOST_ERROR(errorCode, errorText) \
+    PaUtil_SetLastHostErrorInfo(paInDevelopment, errorCode, errorText)
+
+/* PaPulseaudioHostApiRepresentation - host api datastructure specific to this implementation */
+
+#define PULSEAUDIO_TIME_EVENT_USEC 50000
+#define PULSEAUDIO_BUFFER_SIZE (88100 * 4 * 2)
+
+/* Pulseaudio specific functions */
+int PulseaudioCheckConnection(PaPulseaudioHostApiRepresentation *ptr)
+{
+    pa_context_state_t state;
+
+    if (!ptr->context || !ptr->mainloop)
+    {
+        return -1;
+    }
+
+    state = pa_context_get_state(ptr->context);
+
+    if (!PA_CONTEXT_IS_GOOD(state))
+    {
+        return -1;
+    }
+
+    return 0;
+}
+
+PaPulseaudioHostApiRepresentation *PulseaudioNew(void)
+{
+    PaPulseaudioHostApiRepresentation *ptr;
+    int fd[2] = { -1, -1 };
+    char proc[PATH_MAX];
+    char buf[PATH_MAX + 20];
+
+    ptr = (PaPulseaudioHostApiRepresentation*)PaUtil_AllocateMemory(sizeof(PaPulseaudioHostApiRepresentation));
+
+    if (!ptr)
+    {
+        return NULL;
+    }
+
+    ptr->mainloop = pa_threaded_mainloop_new();
+
+    if (!ptr->mainloop)
+    {
+        goto fail;
+    }
+
+    memset(buf, 0x00, PATH_MAX + 20);
+    snprintf(buf, sizeof(buf), "Portaudio Pulseaudio HostApi");
```

Comment:
It would be better to use her the client application name. 
That can be read from for example form 
/proc/4475/cmdline 
or other hacks.